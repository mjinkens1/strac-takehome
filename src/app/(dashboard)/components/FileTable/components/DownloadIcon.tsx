import { ArrowDownTrayIcon, CheckIcon } from "@heroicons/react/24/outline";
import { Tooltip } from "../../../../components/Tooltip";

interface DownloadIconProps {
  downloaded: boolean;
  downloading: boolean;
  progress: number;
}

export function DownloadIcon({
  downloaded,
  downloading,
  progress,
}: DownloadIconProps) {
  if (downloaded) {
    return (
      <CheckIcon data-testid="check-icon" className="size-5 text-green-500" />
    );
  }

  const strokeDasharray = 2 * Math.PI * 16;
  const strokeDashoffset = 2 * Math.PI * 16 * (1 - progress / 100);

  if (downloading) {
    return (
      <div className="relative size-5">
        <svg
          viewBox="0 0 36 36"
          className="absolute top-0 left-0 size-full"
          data-testid="progress-circle"
        >
          <circle
            cx="18"
            cy="18"
            r="16"
            fill="none"
            className="stroke-gray-300 dark:stroke-zinc-700"
            strokeWidth="4"
          />
          <circle
            cx="18"
            cy="18"
            r="16"
            fill="none"
            className="stroke-blue-600 transition-all duration-400"
            strokeWidth="4"
            strokeDasharray={strokeDasharray}
            strokeDashoffset={strokeDashoffset}
            strokeLinecap="round"
          />
        </svg>
      </div>
    );
  }

  return (
    <Tooltip text="Download">
      <ArrowDownTrayIcon data-testid="download-icon" className="size-5" />
    </Tooltip>
  );
}
