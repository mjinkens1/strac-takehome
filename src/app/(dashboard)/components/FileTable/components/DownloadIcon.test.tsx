import { render, screen } from "@testing-library/react";
import { DownloadIcon } from "./DownloadIcon";

// Mock the Tooltip component
jest.mock("../../../../components/Tooltip", () => ({
  Tooltip: ({ children }: { children: React.ReactNode }) => children,
}));

describe("DownloadIcon", () => {
  it("renders check icon when downloaded", () => {
    render(<DownloadIcon downloaded={true} downloading={false} progress={0} />);
    const checkIcon = screen.getByTestId("check-icon");
    expect(checkIcon).toBeInTheDocument();
    expect(checkIcon).toHaveClass("text-green-500");
  });

  it("renders progress circle when downloading", () => {
    const progress = 50;
    const { container } = render(
      <DownloadIcon downloaded={false} downloading={true} progress={progress} />
    );

    const svg = container.querySelector("svg");
    expect(svg).toBeInTheDocument();
    expect(svg).toHaveAttribute("viewBox", "0 0 36 36");

    const circles = container.querySelectorAll("circle");
    expect(circles).toHaveLength(2);

    // Check progress circle properties
    const progressCircle = circles[1];
    expect(progressCircle).toHaveAttribute("strokeLinecap", "round");
    expect(progressCircle).toHaveClass("stroke-blue-600");

    // Verify progress calculation
    const circumference = 2 * Math.PI * 16;
    const expectedOffset = circumference * (1 - progress / 100);
    expect(progressCircle).toHaveAttribute(
      "strokeDashoffset",
      expectedOffset.toString()
    );
  });

  it("renders download icon by default", () => {
    render(
      <DownloadIcon downloaded={false} downloading={false} progress={0} />
    );
    const downloadIcon = screen.getByTestId("download-icon");
    expect(downloadIcon).toBeInTheDocument();
  });

  it("matches snapshot for each state", () => {
    const { container: downloadedContainer } = render(
      <DownloadIcon downloaded={true} downloading={false} progress={0} />
    );
    expect(downloadedContainer).toMatchSnapshot("downloaded");

    const { container: downloadingContainer } = render(
      <DownloadIcon downloaded={false} downloading={true} progress={50} />
    );
    expect(downloadingContainer).toMatchSnapshot("downloading");

    const { container: defaultContainer } = render(
      <DownloadIcon downloaded={false} downloading={false} progress={0} />
    );
    expect(defaultContainer).toMatchSnapshot("default");
  });
});
