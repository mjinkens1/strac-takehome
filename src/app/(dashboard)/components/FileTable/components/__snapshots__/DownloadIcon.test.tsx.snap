// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadIcon matches snapshot for each state: default 1`] = `
<div>
  <svg
    aria-hidden="true"
    class="size-5"
    data-slot="icon"
    data-testid="download-icon"
    fill="none"
    stroke="currentColor"
    stroke-width="1.5"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M3 16.5v2.25A2.25 2.25 0 0 0 5.25 21h13.5A2.25 2.25 0 0 0 21 18.75V16.5M16.5 12 12 16.5m0 0L7.5 12m4.5 4.5V3"
      stroke-linecap="round"
      stroke-linejoin="round"
    />
  </svg>
</div>
`;

exports[`DownloadIcon matches snapshot for each state: downloaded 1`] = `
<div>
  <svg
    aria-hidden="true"
    class="size-5 text-green-500"
    data-slot="icon"
    data-testid="check-icon"
    fill="none"
    stroke="currentColor"
    stroke-width="1.5"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="m4.5 12.75 6 6 9-13.5"
      stroke-linecap="round"
      stroke-linejoin="round"
    />
  </svg>
</div>
`;

exports[`DownloadIcon matches snapshot for each state: downloading 1`] = `
<div>
  <div
    class="relative size-5"
  >
    <svg
      class="absolute top-0 left-0 size-full"
      data-testid="progress-circle"
      viewBox="0 0 36 36"
    >
      <circle
        class="stroke-gray-300 dark:stroke-zinc-700"
        cx="18"
        cy="18"
        fill="none"
        r="16"
        stroke-width="4"
      />
      <circle
        class="stroke-blue-600 transition-all duration-400"
        cx="18"
        cy="18"
        fill="none"
        r="16"
        stroke-dasharray="100.53096491487338"
        stroke-dashoffset="50.26548245743669"
        stroke-linecap="round"
        stroke-width="4"
      />
    </svg>
  </div>
</div>
`;
